lib/src/barback/ddc_transformer.dart

```diff
@@ -0,0 +1,500 @@
+      continue;
+    }
+    var asset = await getInput(id);
+    var contents = await asset.readAsString();
+    var unit = parseDirectives(contents, name: '$id');
+    await _findDependentPackages(
+        basePackage,
+        unit.directives
+            .where((d) => d is UriBasedDirective)
+            .map((d) => _urlToAssetId(
+                asset.id, (d as UriBasedDirective).uri.stringValue, logger))
+            .where((id) => id != null),
+        logger,
+        getInput,
+        hasInput,
+        foundIds: foundIds,
+        foundPackages: foundPackages);
+  }
+  return foundPackages;
+}
+
+File _fileForId(AssetId id, String rootDir, String packagesDir) {
+  return p.url.split(id.path).first == 'lib'
+      ? new File(p.joinAll(
+          [packagesDir, id.package]..addAll(p.url.split(id.path).skip(1))))
+      : new File(p.joinAll([rootDir]..addAll(p.url.split(id.path))));
+}
+
+Set<AssetId> _findSummaryIds(package) {
+  // TODO(jakemac): Read build.yaml if available?
+  return new Set<AssetId>()
+    ..add(
+        new AssetId(package, p.url.join('lib', '$package.$_summaryExtension')));
+}
+
+Future<bool> _isEntryPoint(
+    AssetId id, TransformLogger logger, _InputAsStringReader readInputAsString,
+    {Set<AssetId> seenIds}) async {
+  seenIds ??= new Set<AssetId>();
+  if (!seenIds.add(id)) return false;
+
+  var content = await readInputAsString(id);
+  // Suppress errors at this step, dartdevc will error later if the file can't
+  // parse.
+  var unit = parseCompilationUnit(content, name: '$id', suppressErrors: true);
+  // Logic to check for a valid main(). Currently this is a check for any
+  // top level function called "main" with 2 or fewer parameters. The 2nd
+  // argument is allowed for the isolate conversion test.
+  if (unit.declarations.any((d) =>
+      d is FunctionDeclaration &&
+      d.name.name == 'main' &&
+      d.functionExpression.parameters.parameters.length <= 2)) {
+    return true;
+  }
+
+  // Additionally search all exports that might be exposing a `main` based on
+  // the show/hide settings.
+  var exportsToSearch = unit.directives.where((d) {
+    if (d is! ExportDirective) return false;
+    for (var combinator in (d as ExportDirective).combinators) {
+      if ('${combinator.keyword}' == 'show') {
+        if (combinator.childEntities.any((e) => '$e' == 'main')) return true;
+      } else if ('${combinator.keyword}' == 'hide') {
+        if (combinator.childEntities.any((e) => '$e' == 'main')) return false;
+      }
+    }
+    return true;
+  });
+
+  for (var export in exportsToSearch) {
+    var exportId = _urlToAssetId(
+        id, (export as UriBasedDirective).uri.stringValue, logger);
+    if (exportId == null) continue;
+    if (await _isEntryPoint(exportId, logger, readInputAsString,
+        seenIds: seenIds)) return true;
+  }
+  return false;
+}
+
+AssetId _urlToAssetId(AssetId source, String url, TransformLogger logger) {
```

Comment:
There isn't anything that supports relative imports under the same top level directory though.